query.go

```diff
@@ -442,13 +458,20 @@ func queryChainServiceBatch(
 					break
 				}
 
+				log.Tracef("Query #%v answered, updating state",
+					handleQuery)
+
 				// We got a match signal so we can mark this
 				// query a success.
 				atomic.StoreUint32(&queryStates[handleQuery],
 					uint32(queryAnswered))
 
-				log.Tracef("Query #%v answered, updating state",
-					handleQuery)
+				// If we're done answering all of our queries,
+				// we can exit now.
+				if allDone() {
+					close(allDoneSignal)
```

Comment:
Why not check this in the main loop when we get a response from a goroutine?

Reply:
It’s race-y because we can’t guarantee that the query state has been updated by the time we check it.

Reply:
How is this less racy? In both instances all indexes are checked used atomics. Also it's possible that two goroutines finish at nearly the same time, and cause a panic by double closing the final since they don' tall have their own signal.